File: c++/QuickSort.cpp
#include<bits/stdc++.h>
using namespace std;

int partition(int a[],int s,int e){
    int pivot = a[e];
    int p = s;
    for(int q=s;q<e;q++){
        if(a[q]<pivot){
            swap(a[p],a[q]);
            p++;
        }
    }
    swap(a[p],a[e]);
    return p;
}


void quicksort(int a[],int s,int e){
    if(s>=e) return;

    int p = partition(a,s,e);
    quicksort(a,s,p-1);
    quicksort(a,p+1,e);
}

int main(){
    int n;cin>>n;
    int a[n];
    for(int i=0;i<n;i++){
        cin>>a[i];
    }
    quicksort(a,0,n-1);
    for(int i=0;i<n;i++){
        cout<<a[i]<<" ";
    }
}

File: SortingAndSearching/QuickSort.java
public class Quick_sort {
  
    public static int partition(int arr[],int low,int high){
        int pivot = arr[high];
        int i = low-1;
        for(int j=low;j<high;j++){
            if(arr[j] < pivot){
                i++;
                // swap arr[i] with arr[j]
                int temp=arr[i];
                arr[i]=arr[j];
                arr[j]=temp;
            }
        }
        i++;
        int temp=arr[i];
        arr[i]=arr[high];
        arr[high]=temp;

        return i;
    }
  
    public static void quickSort(int arr[],int low,int high){
        if(low < high){
            int pidx = partition(arr,low,high);
            quickSort(arr,low,pidx-1);
            quickSort(arr,pidx+1,high);
        }
    }
  
    public static void main(String[] args) {
        int[] arr = {6,3,9,5,2,8};
        int n = arr.length;

        quickSort(arr,0,n-1);

        for(int i=0;i<n;i++){
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }
  
}
